### PR TITLE
[CR] Customize Use Item menu and tweak a couple more menus

### PR DESCRIPTION
**Pull request by codemime**
Follows #18996 and #19230.

### Overview

**Use Item menu**

Action names enabled:
act

Action names disabled:
act2

- It's now possible to activate nearby items (like it's done for comestibles and books. See #16787).
- The menu features action names (or `"..." `in case of ambiguity).
- ~~Action names are used for sorting (action name, then item name).~~
- The stuff you can't do anything interesting with (ordinary clothes, for example), isn't shown in the menu.
- Repaired materials are listed in corresponding actions (see cfb2041).
- Added an option to disable action names.

**Read menu**

bookr

- It now shows the same denials.

**Item Action menu**

exec_after

- The menu became shorter: only enabled actions and actions with keybindings are shown.
- It's sorted by action names too.
- Proper column alignment (improved readability).
- No `Cancel` option (exit by ESC).
- You can no longer "turn on" your scarf (specific actions are shown).

How it looked before:
exec_before2
...much more lines...
exec_before3

**Generic UI menu (see above):**

- Colors from selected lines get removed (no gaps in highlighted lines).
- Keybinds are shown for disabled items too (for consistency with the inventory menu).

### Implementation details

- The new `activatable_inventory_preset` for activatables. Both `activatable_inventory_preset` and `read_inventory_preset` inherit from a shared class to share denials.
- Consistency checks for item actions.
- Some actions (unused and duplicated) have been removed, and a lot more missing actions added.
- Gunmod detachment in a separate `iuse` function (see 0fb8cf7).
- `"menu_option_text"` for `"transform"` action (scarfs).
- `item_location::obtain()` was used (which means picking up now consumes moves).

### CODE REVIEW

**Comment by mutt3x** [comment]
First let me say how #18996 and #19230 are really good PRs, i probably dont have to say it. I like those +1 for the book menu.

Now ill be honest about this PR. I personally dont like the use menu. Some may wonder why? Ill try to express why with my limited english.

You see, i just cant seem to comprehend how is this useful, looking at the 1st pic i can asume it would be a visual aid/feature for a NEW player but for someone who is playing this game for a long time its just... why? feels overused.

>Action names are used for sorting

Things that are sorted already (books, drugs, tools) need more sorting? hmm. its not even alpabetical anymore.

I can say its aesthetically better than before since the unused space is ... well used and with colors.
So the main problem for me is the action thing column,
its borderline bloat.
Perhaps an option to enable this. i really dont need to see 'read' on all books. you see what i mean, or smoke on the cigar... its funny.

i just hope you see what im trying to point out. still this is just me maybe other people like this idk.

btw the Item Action menu is good, all fine there. :)

**Comment by Nacarbac** [comment]
Sure, if you've already memorized every item that has a use and what that use is the hard way, it's unnecessary... but that argument can be applied to any quality of life improvement. This is a great improvement in terms of user-friendliness, and being able to active adjacent items is awesome.

File: src/iuse.cpp [hunk 1]
```diff
@@ -6679,6 +6718,10 @@ int iuse::ehandcuffs(player *p, item *it, bool t, const tripoint &pos)
     return it->type->charges_to_use();
 }
 
+/* MACGUFFIN FUNCTIONS
+ * These functions should refer to it->associated_mission for the particulars
+ */
+
```
Comment (Coolthulhu) [inline src/iuse.cpp]: ?

**Comment by Coolthulhu** [comment]
How does it work with unreasonably long descriptions?
In your examples, item names and action descriptions are all sane. What if some of the items had stupidly long names and descriptions? Will the tab get stupidly wide to accommodate the outlier?
If so, it may be worth adding some option to limit tab size. Not necessarily in this PR.

**Comment by harald921** [comment]
> Now ill be honest about this PR. I personally dont like the use menu.

I didn't even know there was a "use item" menu. I feel silly now. I really like this PR though, the UI needs all the polish it can get.

Also, off-topic question - since this PR had a merge conflict, does that mean @codemime will have to rewrite the code in order to fix the error or put up a new PR? I am relatively new to source control and want to be entirely sure of how it all works around here before I start contributing myself.

**Comment by codemime** [comment]
Thanks everyone.

@mutt3x Good points. I'll add an option to disable action names. However, I believe they should be enabled by default. I think you're also right about the sorting, it's excessive.

> What if some of the items had stupidly long names and descriptions? Will the tab get stupidly wide to accommodate the outlier? If so, it may be worth adding some option to limit tab size. 

@Coolthulhu Yes, it will. An option applicable to all columns would be useful. Though, shouldn't long names be generally avoided?

> does that mean @codemime will have to rewrite the code in order to fix the error or put up a new PR? 

I can simply [rebase](https://git-scm.com/book/ch3-6.html) my changes against the existing code. `Git` helps me resolving possible merge conflicts when I do that. Then I [force-push](https://git-scm.com/docs/git-push) them back here. Another (somewhat simpler) alternative is merging the master branch into your branch, but I don't like it, it's crooked.

**Comment by mugling** [comment]
> Good points. I'll add an option to disable action names. However, I believe they should be enabled by default

I can see both viewpoints but we should only implement one

> Another (somewhat simpler) alternative is merging the master branch into your branch, but I don't like it, it's crooked.

It's preferable from a review POV and now we do squash-merges by default is probably not so ugly in the commit history

**Comment by codemime** [comment]
> I can see both viewpoints but we should only implement one

There's no contradiction:  no sorting by action names and an option which is enabled by default.

**Comment by mugling** [comment]
> an option which is enabled by default.

We should avoid configuration switches wherever possible

**Comment by codemime** [comment]
Newbies can benefit from explicit action names alot, but they irritate experienced players. The option IMHO would help.

**Comment by codemime** [comment]
All set.

File: data/json/items/tool_armor.json [hunk 2]
```diff
@@ -1643,7 +1656,8 @@
         "use_action": {
             "type": "transform",
             "msg": "You wrap your scarf a bit tighter.",
-            "target": "scarf_fur"
+            "target": "scarf_fur",
+            "menu_option_text": "Wrap tighter"
```
Comment (mugling) [inline data/json/items/tool_armor.json]: 1. What happens if you omit `menu_option_text`?
2. Field name could be shortened to `menu_text`

**Comment by mugling** [comment]
Needs trivial fix for merge conflict (please don't rebase)

**Comment by mugling** [comment]
Action menu doesn't handle multiple items (for example two holsters)

**Comment by codemime** [comment]
> Action menu doesn't handle multiple items (for example two holsters)

I suppose it never did. I didn't change its behavior, only appearance.

**Comment by codemime** [comment]
> I agree it's not a regression but it could perhaps be resolved here

Probably later. It's not a primary concern of the PR. I've updated the Action menu here for two reasons:
- The number of registered actions has grown so big that the menu required scrolling.
- It showed inconsistent (non-customized) action names.

**Comment by codemime** [comment]
It's ready.

**Comment by saneman226** [comment]
This right here is why I will not go past build 5950 until it is removed or majorly improved.

**Comment by codemime** [comment]
> This right here is why I will not go past build 5950 

So be it!

> until it is removed or majorly improved.

Heh, this obviously means you haven't read the description (and haven't even seen the second screenshot).

**Comment by saneman226** [comment]
The main reason I dislike it is because the window changes form randomly from single column to 3 columns unexpectedly to try to fit my window every time. If we can specify a constraint for these malleable windows in the options, it would be much more consistent in the way it handles the window size as well as how it handles large amounts of items.

Can I also ask why you want to be rude?

**Comment by codemime** [comment]
> Can I also ask why you want to be rude?

I guess my response isn't any more rude than your initial statement. Anyways, I didn't want to be impolite.

> The main reason I dislike it is because the window changes form randomly from single column to 3 columns

This (right here) doesn't change the number of inventory columns at all (I'm referencing the description once again). If you have a good suggestion on this matter, you're free to either post it in #19427 or create a new (non-duplicate) issue. Please remember that "Remove it because I don't like it" - is not a valid reason.

**Pull request merged**